### ISSUE TITLE
Media doesn't appear after import without error message

### ISSUE THREAD

**Issue by githubNutzer89** [gramps-project/gramps-webapi#419]
I don't know if this has anything to do with issue #411 I submitted. After I updated to version 1.2.0 I was able to import media. I tried to import 32 images of which 20 were succesfully imported. The remaining 12 images didn't get imported. Also there was no error message. Neither on the frontend nor in the logs. The images were checked for any checksum missmatch in the database. The zip archive didn't contain any subfolders.

I tried to import the missing images by the grampsweb-sync plugin for gramps on desktop. After checking for any missmatch between server and desktop the plugin showed that 12 images are missing on the server and need to be uploaded. After I hit apply the process failed:

53426: ERROR: grampsapp.py: line 174: Unhandled exception
Traceback (most recent call last):
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapisync.py", line 386, in _async_transfer_media
    self.handle_server_errors(self.upload_files)
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapisync.py", line 393, in handle_server_errors
    return callback(*args)
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapisync.py", line 319, in upload_files
    self.uploaded[gramps_id] = self._upload_file(handle)
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapisync.py", line 331, in _upload_file
    return self.api.upload_media_file(handle=handle, path=path)
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapihandler.py", line 174, in upload_media_file
    self._upload_file(url=url, fobj=f)
  File "C:\Users\Christoph\AppData\Roaming\gramps\gramps51\plugins\GrampsWebSync\webapihandler.py", line 190, in _upload_file
    with urlopen(req) as res:
  File "AIO/urllib/request.py", line 223, in urlopen
  File "AIO/urllib/request.py", line 526, in open
  File "AIO/urllib/request.py", line 544, in _open
  File "AIO/urllib/request.py", line 504, in _call_chain
  File "AIO/urllib/request.py", line 1346, in http_open
  File "AIO/urllib/request.py", line 1321, in do_open
  File "AIO/http/client.py", line 1331, in getresponse
  File "AIO/http/client.py", line 297, in begin
  File "AIO/http/client.py", line 266, in _read_status
http.client.RemoteDisconnected: Remote end closed connection without response

What could be the reason?

**Comment by DavidMStraub** [comment]
Hi, concerning the media upload part, can you share the log of the celery worker when you uploaded the 32 files?

**Comment by githubNutzer89** [comment]
What do you mean by "celery worker logs"? The normal container log? It doesn't show anything at the time of the import.

**Comment by DavidMStraub** [comment]
Please share more details about your setup. If you use the compose from the documentation, it should be `docker-compose logs grampsweb_celery`.

**Comment by githubNutzer89** [comment]
When I execute this command it says "No such service: grampsweb_celery". I executed the command in the respective folder where my docker_compose.yml lives in.

**Comment by DavidMStraub** [comment]
Please share your redacted docker-compose.yml

**Comment by githubNutzer89** [comment]
version: "3.7"

services:
  grampsweb:
    container_name: grampsweb
    image: ghcr.io/gramps-project/grampsweb:latest
    restart: unless-stopped
    ports:
      - "5054:5000"  # host:docker
    environment:
      TREE: "Gramps Web"  # will create a new tree if not exists
    volumes:
      - gramps_users:/app/users  # persist user database
      - gramps_index:/app/indexdir  # persist search index
      - gramps_thumb_cache:/app/thumbnail_cache  # persist thumbnails
      - gramps_secret:/app/secret  # persist flask secret
      - /volume1/docker/grampsweb/grampsdb:/root/.gramps/grampsdb  # persist Gramps database
      - /volume1/docker/grampsweb/gramps_media:/app/media  # persist media files

volumes:
  gramps_users:
  gramps_index:
  gramps_thumb_cache:
  gramps_secret:
  gramps_db:
  gramps_media:

**Comment by DavidMStraub** [comment]
I see, no celery. Then it could be that the task just times out. How long did it take?

**Comment by githubNutzer89** [comment]
Ah, I am sorry. Now I know, what you meant. 

The celery log:

2023-08-16T17:12:16.447100706Z [2023-08-16 17:12:16,446: INFO/MainProcess] Task gramps_webapi.api.tasks.import_media_archive[f1de2bed-0b7c-4e7d-9015-bb46bbc0cbeb] received

2023-08-16T17:12:16.482676514Z [2023-08-16 17:12:16,482: INFO/ForkPoolWorker-1] Task gramps_webapi.api.tasks.import_media_archive[f1de2bed-0b7c-4e7d-9015-bb46bbc0cbeb] succeeded in 0.032827794551849365s: {'missing': 0, 'uploaded': 0, 'failures': 0}

**Comment by DavidMStraub** [comment]
Ok, I assume that's not the original upload, right? So now it says nothing is missing. But you say some are not uploaded? Are they missing on the server? Could it be that the have absolute paths?

**Comment by githubNutzer89** [comment]
Disregard the upper logs. Something went wrong with the container after I redeployed everything. So I had a clean database, imported the xml from Gramps Desktop and after that I imported the media (all images, no subfolders within the zip archive). Before I imported anything I checked that all pictures had relative paths and let Gramps Desktop check if there is any discrepancies within my database, e.g. checksum. 

Now I got this log:

[2023-08-16 17:27:35,396: INFO/MainProcess] Task gramps_webapi.api.tasks.import_media_archive[7d5ac285-1729-440c-a561-61565b675d5d] received

[2023-08-16 17:27:37,160: INFO/ForkPoolWorker-1] Task gramps_webapi.api.tasks.import_media_archive[7d5ac285-1729-440c-a561-61565b675d5d] succeeded in 1.761456111446023s: {'missing': 32, 'uploaded': 20, 'failures': 12}

**Comment by DavidMStraub** [comment]
Since you're using local media files, pretty much the only error I can think of is this https://github.com/gramps-project/gramps-webapi/blob/master/gramps_webapi/api/media.py#L124-L126, thus my question about absolute paths.

**Comment by githubNutzer89** [comment]
All pathes are relative. I doublechecked it and I also used the respective gramplet to convert absolute paths into relatives. But there is on thing all failed images have in common. They live in subfolders of my Gramps Desktop media folder. So for the sake of simplicity let's say my base directory is c:\User\...\media. All media files which are plain in this folder have been succesfully imported. All others which are organised in subfolders failed like c:\User\...\media\bad quality\... . The imported zip archive doesn't have the subfolder structure.

**Comment by DavidMStraub** [comment]
Ok, that's an interesting finding thanks. Will investigate what could cause that :detective:

**Comment by DavidMStraub** [comment]
Actually - probably this is just #187?

**Comment by DavidMStraub** [comment]
I still don't get it - `bad quality\grandfather.jpg` is a perfectly valid Posix filename, so why would that raise an error... we should really add more debug logging.

**Comment by githubNutzer89** [comment]
Alright, I rearranged all media files on my local Gramps Desktop machine to be plain in the media base folder without any subfolders and it worked. Now all files are recognized. 

Maybe a space within the path can cause the error?

Btw, what happens if I move media files on my local machine, fix the paths in Gramps Desktop and sync with gramps web? Can I run in any issues there? What triggers a change for the syncing process?

**Comment by DavidMStraub** [comment]
> Maybe a space within the path can cause the error since linux interpretes a space as a separator for commands?

No, it shouldn't - I think I have plenty of spaces in paths myself. If not having subfolders makes it work, I am pretty sure it has to do with Windows backslashes - #187.

> Btw, what happens if I move media files on my local machine, fix the paths in Gramps Desktop and sync with gramps web? Can I run in any issues there? What triggers a change for the syncing process?

It depends on your setup. If you use [object storage for media](https://www.grampsweb.org/s3/), like Grampshub does, it doesn't matter - it addresses media files by their checksum and the path in the Gramps database doesn't matter at all, even when it changes.

If you use local media and change the path, the files will disappear on Gramps Web and you will have to re-upload them.

**Comment by elnigno** [comment]
I ran into a similar problem and might have found the cause. I imported my family tree from Gramps Desktop to Gramps Web (running on my NAS with docker), following the instructions here: https://www.grampsweb.org/user-guide/import/
Importing genealogical data worked, but media import with a zip file failed. I also tried the Gramps Web Sync addon, but it would fails at the media sync step. I checked the logs for the `grampsweb` container, and found this stack trace (I've just redacted the image path at the end):

```
Traceback (most recent call last):
File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 135, in handle
  self.handle_request(listener, req, client, addr)
File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 193, in handle_request
  util.reraise(*sys.exc_info())
File "/usr/local/lib/python3.9/dist-packages/gunicorn/util.py", line 641, in reraise
  raise value
File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 178, in handle_request
  respiter = self.wsgi(environ, resp.start_response)
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2213, in __call__
  return self.wsgi_app(environ, start_response)
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2193, in wsgi_app
  response = self.handle_exception(e)
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2190, in wsgi_app
  response = self.full_dispatch_request()
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1486, in full_dispatch_request
  rv = self.handle_user_exception(e)
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1484, in full_dispatch_request
  rv = self.dispatch_request()
File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1469, in dispatch_request
  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/auth.py", line 44, in wrapper
  return func(*args, **kwargs)
File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 109, in view
  return current_app.ensure_sync(self.dispatch_request)(**kwargs)
File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 190, in dispatch_request
  return current_app.ensure_sync(meth)(**kwargs)
File "/usr/local/lib/python3.9/dist-packages/webargs/core.py", line 649, in wrapper
  return func(*args, **kwargs)
File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/file.py", line 107, in put
  media_handler.upload_file(f, checksum, mime, path=path)
File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/media.py", line 129, in upload_file
  upload_file_local(self.base_dir, rel_path, stream)
File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/file.py", line 222, in upload_file_local
  fs.save(os.path.join(base_dir, rel_path))
File "/usr/local/lib/python3.9/dist-packages/werkzeug/datastructures/file_storage.py", line 125, in save
  dst = open(dst, "wb")
FileNotFoundError: [Errno 2] No such file or directory: '/app/media/folder1/folder2/folder3/image.jpg'
```

I opened a console into the `grampsweb` container, navigated to the `/app/media` directory and found it empty. So I tried recreating the folder structure with

```bash
mkdir -p /app/media/folder1/folder2/folder3
```

I tried running again Gramps Web Sync and it still failed, but it actually partially succeeded because I could now see `image.jpg` in Gramps Web.

So it seems to me that the problem here is that the media file import code assumed that the folders containing the file it's importing exist. When that is not the case it fails without attempting to create them.

I hope this is useful.

P.S.: I think there's a bug when changing the path of an image in Gramps Desktop, then syncing to Gramps Web (assuming they were in sync before that change); in Gramps Web the image will be synchronized to the new path (in the container's media folder), but the image in the old path will not be deleted. So it seems orphan media is created. If you think this is a bug, I can try to conjure a more formal repro and open a bug.

**Comment by DavidMStraub** [comment]
Thanks @elnigno, you'r absolutely right! `FileStorage.save` doesn't create missing subdirectories and it went unnoticed so far because it didn't happen in the test cases :unamused:  In my own deployments I didn't notice because I use the object storage file backend.

Concerning the PS: I wouldn't call it a bug, but a known limitation. In fact, currently Gramps does not delete _any_ media files. One of the reasons is that many people have their Gramps media files in directories with lots of other files (that Gramps doesn't know about). I think it's too dangerous deleting files just because the path (or the checksum) does not appear in the Gramps database.

For bigger deployments, I would recommend using the object storage file backend; in this case, files are stored according to their checksum, not their path in the database, so changing the path does not require moving/deleting/reuploading files.

**Comment by elnigno** [comment]
Glad I could help!

I see why one would take a conservative approach when deleting files. Perhaps I can write a script to find orphaned files in the media folder, that are not referenced in gramps.

**Comment by DavidMStraub** [comment]
By the way, the behaviour of Web API is in line with Gramps desktop - Gramps desktop also never deletes media files, even when you delete a media object. It has the [Media Verify](https://www.gramps-project.org/wiki/index.php/Addon:Media_Verify_Tool) to detect mismatches between files and database.

**Comment by DavidMStraub** [comment]
Fixed by #426 - please reopen if the problem persists with the next release.
